File: ros_packages/autopilot_cpp/src/motorboat_autopilot_node.cpp
#include "motorboat_autopilot_node.hpp"


using namespace std::chrono_literals;
using std::placeholders::_1;



MotorboatAutopilotNode::MotorboatAutopilotNode() : Node("motorboat_autopilot_cpp") {

    std::string package_share_directory = ament_index_cpp::get_package_share_directory("autopilot_cpp");
    std::string json_file_path = package_share_directory + "/config/motorboat_default_parameters.json";

    std::ifstream file(json_file_path);
    json config = json::parse(file);

    // create a dictionary directly out of the default values of the autopilot parameters
    for (auto it = config.begin(); it != config.end(); ++it) {
        autopilot_parameters[it.key()] = it.value()["default"];
    }

    auto transient_qos = rclcpp::QoS(1).reliable().transient_local();
    config_path_publisher = this->create_publisher<std_msgs::msg::String>("/autopilot_param_config_path", transient_qos);

    config_path_publisher->publish(std_msgs::msg::String().set__data(json_file_path));


    motorboat_autopilot = MotorboatAutopilot(&autopilot_parameters);

    rclcpp::SensorDataQoS sensor_qos = rclcpp::SensorDataQoS();

    // Timer
    autopilot_refresh_timer = create_wall_timer(
        std::chrono::duration<float>(1.0 / autopilot_parameters["autopilot_refresh_rate"].get<float>()),
        std::bind(&MotorboatAutopilotNode::update_ros_topics, this)
    );


    // Subscriptions
    sub.push_back(create_subscription<std_msgs::msg::String>("/autopilot_parameters", 10, std::bind(&MotorboatAutopilotNode::autopilot_parameters_callback, this, _1)));
    sub.push_back(create_subscription<autoboat_msgs::msg::WaypointList>("/waypoints_list", 10, std::bind(&MotorboatAutopilotNode::waypoints_list_callback, this, _1)));

    sub.push_back(create_subscription<sensor_msgs::msg::NavSatFix>("/position", 10, std::bind(&MotorboatAutopilotNode::position_callback, this, _1)));
    sub.push_back(create_subscription<std_msgs::msg::Float32>("/heading", 10, std::bind(&MotorboatAutopilotNode::heading_callback, this, _1)));

    sub.push_back(create_subscription<geometry_msgs::msg::Twist>("/velocity", 10, std::bind(&MotorboatAutopilotNode::velocity_callback, this, _1)));
    sub.push_back(create_subscription<autoboat_msgs::msg::RCData>("/rc_data", sensor_qos, std::bind(&MotorboatAutopilotNode::rc_data_callback, this, _1)));
    sub.push_back(create_subscription<std_msgs::msg::Bool>("/object_detection_emergency_stop", 10, std::bind(&MotorboatAutopilotNode::emergency_stop_callback, this, _1)));



    // Publishers
    current_waypoint_index_publisher = this->create_publisher<std_msgs::msg::Int32>("/current_waypoint_index", 10);
    autopilot_mode_publisher = this->create_publisher<std_msgs::msg::UInt8>("/autopilot_mode", sensor_qos);
    full_autonomy_maneuver_publisher = this->create_publisher<std_msgs::msg::UInt8>("/full_autonomy_maneuver", sensor_qos);

    should_propeller_motor_be_powered_publisher = create_publisher<std_msgs::msg::Bool>("/should_propeller_motor_be_powered", 10);
    propeller_motor_control_struct_publisher = create_publisher<autoboat_msgs::msg::VESCControlData>("/propeller_motor_control_struct", 10);

    desired_rudder_angle_publisher = create_publisher<std_msgs::msg::Float32>("/desired_rudder_angle", 10);
    zero_rudder_encoder_publisher = create_publisher<std_msgs::msg::Bool>("/zero_rudder_encoder", 10);

    // this is not really for actually controlling the boat but more for debugging
    desired_heading_publisher = this->create_publisher<std_msgs::msg::Float32>("/desired_heading", 10);
}




void MotorboatAutopilotNode::position_callback(const sensor_msgs::msg::NavSatFix::SharedPtr msg) {
    current_latitude = msg->latitude;
    current_longitude = msg->longitude;
    last_gps_position_received_time = this->now();
}

void MotorboatAutopilotNode::velocity_callback(const geometry_msgs::msg::Twist::SharedPtr msg) {
    current_global_velocity = std::vector<float>({static_cast<float>(msg->linear.x), static_cast<float>(msg->linear.y)});
    current_speed = std::hypot(msg->linear.x, msg->linear.y);
}

void MotorboatAutopilotNode::heading_callback(const std_msgs::msg::Float32::SharedPtr msg) {
    current_heading = msg->data;
    last_heading_received_time = this->now();
}



void MotorboatAutopilotNode::rc_data_callback(const autoboat_msgs::msg::RCData::SharedPtr msg) {
    last_rc_data_time = this->now();

    // Hold heading logic
    if (msg->toggle_f == 1 && previous_toggle_f != 1) {
        heading_to_hold = current_heading;
    }
    previous_toggle_f = msg->toggle_f;

    // Rudder Zeroing Logic
    if (!previous_button_d && msg->button_d) {
        should_zero_encoder = true;
        encoder_has_been_zeroed = false;
    }
    else if (encoder_has_been_zeroed) {
        should_zero_encoder = false;
    }

    previous_button_d = msg->button_d;

    joystick_left_y = -1.0 * msg->joystick_left_y;
    joystick_right_x = msg->joystick_right_x;

    // Mode Logic
    if (msg->toggle_b == 1) should_propeller_motor_be_powered = false;
    else should_propeller_motor_be_powered = true;

    if (msg->toggle_b == 2) autopilot_mode = MotorboatControlModes::DISABLED;
    else if (msg->toggle_f == 2) autopilot_mode = MotorboatControlModes::WAYPOINT_MISSION;
    else if (msg->toggle_f == 1) autopilot_mode = MotorboatControlModes::HOLD_HEADING;
    else autopilot_mode = MotorboatControlModes::FULL_RC;

    // Control Type
    if (msg->toggle_c == 0) propeller_motor_control_mode = PropellerMotorControlMode::RPM;
    else if (msg->toggle_c == 1) propeller_motor_control_mode = PropellerMotorControlMode::DUTY_CYCLE;
    else propeller_motor_control_mode = PropellerMotorControlMode::CURRENT;
}



void MotorboatAutopilotNode::autopilot_parameters_callback(const std_msgs::msg::String::SharedPtr new_parameters) {

    json new_parameters_json = json::parse(new_parameters->data);

    for (auto it = new_parameters_json.begin(); it != new_parameters_json.end(); ++it) {
        std::string key = it.key();

        if (autopilot_parameters.find(key) == autopilot_parameters.end()) {
            RCLCPP_INFO(this->get_logger(), "New parameter received: %s", key.c_str());
        }

        if (it.value().is_object() && it.value().contains("default")) {
            autopilot_parameters[key] = it.value()["default"];
        }
        else {
            autopilot_parameters[key] = it.value();
        }
    }


    // Handle autopilot refresh rate differently, since it isn't a simple value
    if (new_parameters_json.contains("autopilot_refresh_rate")) {
        RCLCPP_INFO(this->get_logger(), "Updating autopilot refresh rate...");

        if (autopilot_refresh_timer) autopilot_refresh_timer->cancel();
        float period = 1.0 / autopilot_parameters["autopilot_refresh_rate"].get<float>();
        autopilot_refresh_timer = this->create_wall_timer(std::chrono::duration<float>(period), std::bind(&MotorboatAutopilotNode::update_ros_topics, this));
    }
}


void MotorboatAutopilotNode::waypoints_list_callback(const autoboat_msgs::msg::WaypointList::SharedPtr waypoint_list) {
    if (waypoint_list->waypoints.empty()) {
        return;
    }

    // Reset the internal autopilot state
    motorboat_autopilot.reset();

    std::vector<Position> waypoint_positions;

    // Convert each NavSatFix to our internal Position class
    for (const auto & navsat : waypoint_list->waypoints) {
        waypoint_positions.emplace_back(navsat.longitude, navsat.latitude);
    }

    // Pass the vector to the autopilot logic
    motorboat_autopilot.update_waypoints_list(waypoint_positions);

    RCLCPP_INFO(this->get_logger(), "Received %zu new waypoints.", waypoint_positions.size());
}




void MotorboatAutopilotNode::emergency_stop_callback(const std_msgs::msg::Bool::SharedPtr msg) {
    if (msg->data && autopilot_mode == MotorboatControlModes::WAYPOINT_MISSION) {
        autopilot_mode = MotorboatControlModes::EMERGENCY_STOP;
    } else if (!msg->data) {
        autopilot_mode = MotorboatControlModes::WAYPOINT_MISSION;
    }
}


void MotorboatAutopilotNode::update_ros_topics() {
    std::optional<float> desired_rudder_angle = 0.0f;
    std::string desired_vesc_control_type = "rpm";
    float desired_vesc_control_value = 0.0f;

    // Disconnect Safety Check
    auto now = this->get_clock()->now();
    double rc_dt = (now - last_rc_data_time).seconds();
    double gps_dt = (now - last_gps_position_received_time).seconds();
    double heading_dt = (now - last_heading_received_time).seconds();

    bool has_rc_disconnected = rc_dt >= autopilot_parameters["rc_data_failsafe_time"].get<double>();
    bool has_gps_disconnected = gps_dt >= autopilot_parameters["gps_position_failsafe_time"].get<double>();
    bool has_heading_disconnected = heading_dt >= autopilot_parameters["heading_data_failsafe_time"].get<double>();
    bool has_sensor_disconnected = has_rc_disconnected || has_gps_disconnected || has_heading_disconnected;

    if (has_sensor_disconnected) {
        RCLCPP_INFO(this->get_logger(), "has sensor disconnected: %d", has_sensor_disconnected);
        RCLCPP_INFO(this->get_logger(), "has remote controller disconnected: %d", has_rc_disconnected);
        RCLCPP_INFO(this->get_logger(), "has gps disconnected: %d", has_gps_disconnected);
        RCLCPP_INFO(this->get_logger(), "has heading disconnected: %d", has_heading_disconnected);

        desired_vesc_control_type = "rpm";
        desired_vesc_control_value = 0.0f;
        desired_rudder_angle = 0.0f;
    }
    else if (autopilot_mode == MotorboatControlModes::EMERGENCY_STOP) {
        desired_vesc_control_type = "rpm";
        desired_vesc_control_value = 0.0f;
        desired_rudder_angle = autopilot_parameters["rudder_hard_over"].get<float>();
    }
    else if (autopilot_mode == MotorboatControlModes::WAYPOINT_MISSION) {
        Position pos(current_longitude, current_latitude);
        auto [desired_rpm, desired_rudder] = motorboat_autopilot.run_waypoint_mission_step(pos, current_heading);
        desired_vesc_control_type = "rpm";
        desired_vesc_control_value = desired_rpm;
        desired_rudder_angle = desired_rudder;
    }
    else if (autopilot_mode == MotorboatControlModes::HOLD_HEADING) {
        desired_rudder_angle = motorboat_autopilot.get_optimal_rudder_angle(current_heading, heading_to_hold);
        auto [vesc_type, vesc_val, _] = motorboat_autopilot.run_rc_control(
            joystick_left_y, joystick_right_x, propeller_motor_control_mode
        );
        desired_vesc_control_type = vesc_type;
        desired_vesc_control_value = vesc_val;
    }
    else if (autopilot_mode == MotorboatControlModes::FULL_RC) {
        auto [vesc_type, vesc_val, desired_rudder] = motorboat_autopilot.run_rc_control(
            joystick_left_y, joystick_right_x, propeller_motor_control_mode
        );
        desired_vesc_control_type = vesc_type;
        desired_vesc_control_value = vesc_val;
        desired_rudder_angle = desired_rudder;
    }
    else {
        desired_vesc_control_type = "rpm";
        desired_vesc_control_value = 0.0f;
        desired_rudder_angle = 0.0f;
    }

    // Now publish to respective topics
    propeller_motor_control_struct_publisher->publish(autoboat_msgs::msg::VESCControlData()
        .set__control_type_for_vesc(desired_vesc_control_type)
        .set__control_value(desired_vesc_control_value)
    );

    if (desired_rudder_angle.has_value()) {
        desired_rudder_angle_publisher->publish(std_msgs::msg::Float32().set__data(*desired_rudder_angle));
    }

    if (should_zero_encoder) {
        zero_rudder_encoder_publisher->publish(std_msgs::msg::Bool().set__data(true));
        encoder_has_been_zeroed = true;
        should_zero_encoder = false; // Reset flag after publishing
    }

    should_propeller_motor_be_powered_publisher->publish(std_msgs::msg::Bool().set__data(should_propeller_motor_be_powered));

    current_waypoint_index_publisher->publish(std_msgs::msg::Int32().set__data(motorboat_autopilot.get_current_waypoint_index()));

    autopilot_mode_publisher->publish(std_msgs::msg::UInt8().set__data(static_cast<uint8_t>(autopilot_mode)));

    full_autonomy_maneuver_publisher->publish(std_msgs::msg::UInt8().set__data(255));

    if (autopilot_mode == MotorboatControlModes::HOLD_HEADING) {
        desired_heading_publisher->publish(std_msgs::msg::Float32().set__data(heading_to_hold));
    }
    else if (autopilot_mode == MotorboatControlModes::WAYPOINT_MISSION && motorboat_autopilot.get_current_waypoints_list().size() > 0) {
        Position pos(current_longitude, current_latitude);
        int waypoint_index = motorboat_autopilot.get_current_waypoint_index();
        auto waypoints_list = motorboat_autopilot.get_current_waypoints_list();

        if (waypoint_index >= 0 && waypoint_index < (int)waypoints_list.size()) {
            float bearing = get_bearing(pos, waypoints_list[waypoint_index]);
            desired_heading_publisher->publish(std_msgs::msg::Float32().set__data(bearing));
        }
        else {
            desired_heading_publisher->publish(std_msgs::msg::Float32().set__data(0.0f));
        }
    }
    else {
        desired_heading_publisher->publish(std_msgs::msg::Float32().set__data(0.0f));
    }
}


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MotorboatAutopilotNode>());
    rclcpp::shutdown();
    return 0;
}

File: ros_packages/autopilot/autopilot/motorboat_autopilot_node.py
import json
import time
from typing import Any

import numpy as np
import numpy.typing as npt
import rclpy
from autoboat_msgs.msg import RCData, VESCControlData, WaypointList
from geometry_msgs.msg import Twist
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import NavSatFix
from std_msgs.msg import Bool, Float32, Int32, String

from .autopilot_library.motorboat_autopilot import MotorboatAutopilot
from .autopilot_library.utils.constants import (
    CONFIG_DIRECTORY,
    QOS_AUTOPILOT_PARAMETER_CONFIG_PATH,
    MotorboatControlModes,
    PropellerMotorControlMode,
)
from .autopilot_library.utils.position import Position
from .autopilot_library.utils.utils_function_library import get_bearing


class MotorboatAutopilotNode(Node):
    """
    The autopilot takes in a bunch of sensor data and waypoints and attempts to traverses through
    the waypoints by continuously publishing to the propeller motor control struct and rudder angle topics.

    The main function that you should pay attention to is ``update_ros_topics``,
    since this is the function that is called periodically on a timer.

    NOTE: All units are in standard SI units and angles are generally measured in degrees unless otherwise specified.
    """

    def __init__(self) -> None:
        super().__init__("motorboat_autopilot")

        self.logger = self.get_logger()

        self.autopilot_parameter_config_path_publisher = self.create_publisher(
            String, "/autopilot_parameter_config_path", QOS_AUTOPILOT_PARAMETER_CONFIG_PATH
        )
        parameters_path = CONFIG_DIRECTORY / "motorboat_default_parameters.json"
        self.autopilot_parameter_config_path_publisher.publish(String(data=parameters_path.as_posix()))

        with open(parameters_path, "r", encoding="utf-8") as parameters_file:
            self.raw_autopilot_parameters: dict[str, dict[str, Any]] = json.load(parameters_file)

        # structured like {parameter_name: parameter_value}
        self.autopilot_parameters: dict[str, Any] = {}
        for param in self.raw_autopilot_parameters:
            self.autopilot_parameters[param] = self.raw_autopilot_parameters[param]["default"]


        self.motorboat_autopilot = MotorboatAutopilot(parameters=self.autopilot_parameters, logger=self.get_logger())

        # Initialize ROS2 subscriptions, publishers, and timers
        autopilot_refresh_period = 1 / self.autopilot_parameters["autopilot_refresh_rate"]
        self.autopilot_refresh_timer = self.create_timer(autopilot_refresh_period, self.update_ros_topics)

        self.create_subscription(String, "/autopilot_parameters", self.autopilot_parameters_callback, 10)
        self.create_subscription(WaypointList, "/waypoints_list", self.waypoints_list_callback, 10)
        self.create_subscription(NavSatFix, "/position", self.position_callback, qos_profile_sensor_data)
        self.create_subscription(Twist, "/velocity", self.velocity_callback, qos_profile_sensor_data)
        self.create_subscription(Float32, "/heading", self.heading_callback, qos_profile_sensor_data)
        self.create_subscription(RCData, "/rc_data", self.rc_data_callback, qos_profile_sensor_data)
        self.create_subscription(Bool, "/object_detection_emergency_stop", self.emergency_stop_callback, 10)

        self.current_waypoint_index_publisher = self.create_publisher(Int32, "/current_waypoint_index", 10)
        self.boat_control_mode_publisher = self.create_publisher(String, "/boat_control_mode", qos_profile_sensor_data)
        self.desired_heading_publisher = self.create_publisher(Float32, "/desired_heading", 10)

        self.should_propeller_motor_be_powered_publisher = self.create_publisher(Bool, "/should_propeller_motor_be_powered", 10)
        self.propeller_motor_control_struct_publisher = self.create_publisher(
            VESCControlData, "/propeller_motor_control_struct", qos_profile_sensor_data
        )

        self.desired_rudder_angle_publisher = self.create_publisher(Float32, "/desired_rudder_angle", 10)
        self.zero_rudder_encoder_publisher = self.create_publisher(Bool, "/zero_rudder_encoder", 10)





        # default values
        self.position = Position(longitude=0.0,latitude=0.0)
        self.velocity: npt.NDArray[np.float64] = np.zeros(2, dtype=np.float64)
        self.speed = 0.0
        self.heading = 0.0
        self.rudder_angle = 0.0

        self.motorboat_control_mode = MotorboatControlModes.WAYPOINT_MISSION
        self.propeller_motor_control_mode = PropellerMotorControlMode.RPM
        self.should_propeller_motor_be_powered = False
        self.should_zero_encoder = False
        self.encoder_has_been_zeroed = False
        self.heading_to_hold = 0.0
        self.has_default_autopilot_parameters_been_received_by_telemetry_node = False

        # used to check whether we have disconnected from any of the sensors and stop the propeller
        self.last_rc_data_received_time: float = 0.0
        self.last_gps_position_received_time: float = 0.0
        self.last_heading_received_time: float = 0.0

        self.joystick_left_x: float = 0.0
        self.joystick_left_y: float = 0.0
        self.joystick_right_x: float = 0.0
        self.joystick_right_y: float = 0.0

        self.button_a: bool = False
        self.toggle_b: int = 0
        self.toggle_c: int = 0
        self.button_d: bool = False
        self.toggle_e: int = 0
        self.toggle_f: int = 0




    def rc_data_callback(self, rc_data_message: RCData) -> None:
        """
        This callback is called whenever there is new data about what is being pressed on the remote control.
        Whenever this callback is called, a couple of things happen:
        1. We check whether or not we are trying to zero the rudder and if we have, then we should handle that.

        2. We need to keep track of if we are entering hold heading mode, and if we are,
        then we need to keep track of the current heading, since that heading is the one we will need to hold.

        3. We set the SailboatControlMode based on a specific combination of toggles to determine whether we are in
        full rc mode, full autopilot mode, or one of the semi autonomous modes.
        Finally, we need to set the propeller control mode based on the toggles to determine whether we are controlling
        the propeller based on current output, duty cycle output, or propeller RPM.

        Parameters
        ----------
            rc_data_message (RCData): A struct that contains all of the data on what is pressed on the remote control
        """

        self.last_rc_data_received_time = time.time()

        # This means we have entered hold heading mode, so keep track of the current heading since this is the target heading
        if rc_data_message.toggle_f == 1 and self.toggle_f != 1:
            self.heading_to_hold = self.heading

        # Are we trying to zero the rudder?
        if self.button_d is False and rc_data_message.button_d is True:
            self.should_zero_encoder = True
            self.encoder_has_been_zeroed = False

        elif self.encoder_has_been_zeroed:
            self.should_zero_encoder = False

        self.joystick_left_x = rc_data_message.joystick_left_x

        # this is negated because positive throttle ends up making the boat go in reverse
        self.joystick_left_y = -1 * rc_data_message.joystick_left_y

        self.joystick_right_x = rc_data_message.joystick_right_x
        self.joystick_right_y = rc_data_message.joystick_right_y

        self.button_a = rc_data_message.button_a
        self.toggle_b = rc_data_message.toggle_b
        self.toggle_c = rc_data_message.toggle_c
        self.button_d = rc_data_message.button_d
        self.toggle_e = rc_data_message.toggle_e
        self.toggle_f = rc_data_message.toggle_f

        # kill switch
        if self.toggle_b == 0:
            self.should_propeller_motor_be_powered = True
        elif self.toggle_b == 1:
            self.should_propeller_motor_be_powered = False

        if self.toggle_b == 2:
            self.motorboat_control_mode = MotorboatControlModes.DISABLED

        # full autonomy
        elif self.toggle_f == 2:
            self.motorboat_control_mode = MotorboatControlModes.WAYPOINT_MISSION

        # hold heading to the direction that we started this mode in
        elif self.toggle_f == 1:
            self.motorboat_control_mode = MotorboatControlModes.HOLD_HEADING

        # remote controlled
        elif self.toggle_f == 0:
            self.motorboat_control_mode = MotorboatControlModes.FULL_RC
        # should not happen
        else:
            print("WARNING: INCORRECT COMBINATION OF RC SWITCHES USED")


        # Check the type of control that is selected
        if self.toggle_c == 0:
            self.propeller_motor_control_mode = PropellerMotorControlMode.RPM

        elif self.toggle_c == 1:
            self.propeller_motor_control_mode = PropellerMotorControlMode.DUTY_CYCLE

        elif self.toggle_c == 2:
            self.propeller_motor_control_mode = PropellerMotorControlMode.CURRENT




    def autopilot_parameters_callback(self, new_parameters: String) -> None:
        """
        Updates the autopilot parameters based on the input JSON string.

        NOTE: Only parameters that already exist in the autopilot will be updated.

        Parameters
        ----------
        new_parameters
            A JSON string that contains the new parameters to set for the autopilot.
        """

        new_parameters_json: dict = json.loads(new_parameters.data)

        for new_parameter_name, new_parameter_value in new_parameters_json.items():
            if new_parameter_name not in self.autopilot_parameters:
                print("WARNING: Attempted to set an autopilot parameter that the autopilot doesn't know")
                print("If you would like to make a new autopilot parameter, please edit motorboat_default_parameters.json")
                continue

            self.autopilot_parameters[new_parameter_name] = new_parameter_value

        # special cases to handle since they do not update automatically
        if "autopilot_refresh_rate" in new_parameters_json:
            self.destroy_timer(self.autopilot_refresh_timer)
            self.autopilot_refresh_timer = self.create_timer(
                1 / self.autopilot_parameters["autopilot_refresh_rate"], self.update_ros_topics
            )


    def waypoints_list_callback(self, waypoint_list: WaypointList) -> None:
        """
        Callback function that is called whenever there is a new waypoint list message.
        Converts the list of ROS2 ``NavSatFix`` objects to a list of ``Position`` objects,
        which are a custom datatype that has some useful helper methods.

        Does not directly set the waypoints in the motorboat autopilot object,
        instead it calls the ``update_waypoints_list`` method of the motorboat autopilot object.

        Parameters
        ----------
        waypoint_list
            A ROS2 message that contains a list of waypoints as ``NavSatFix`` objects.
        """

        if len(waypoint_list.waypoints) == 0:
            return

        self.motorboat_autopilot.reset()

        waypoint_navsatfixes: list[NavSatFix] = waypoint_list.waypoints

        waypoint_positions: list[Position] = []
        for waypoint in waypoint_navsatfixes:
            waypoint_positions.append(Position(longitude=waypoint.longitude, latitude=waypoint.latitude))

        self.motorboat_autopilot.update_waypoints_list(waypoint_positions)


    def position_callback(self, position: NavSatFix) -> None:
        """A callback function to get the current position of the boat."""
        self.position = Position(longitude=position.longitude, latitude=position.latitude)
        self.last_gps_position_received_time = time.time()


    def velocity_callback(self, velocity: Twist) -> None:
        """A callback function to get the current velocity and speed of the boat."""
        self.velocity = np.array([velocity.linear.x, velocity.linear.y], dtype=np.float64)
        self.speed = np.linalg.norm(self.velocity)


    def heading_callback(self, heading: Float32) -> None:
        """A callback function to get the current heading of the boat."""
        self.heading = heading.data
        self.last_heading_received_time = time.time()

    def emergency_stop_callback(self, should_emergency_stop: Bool)-> None:
        """
        A callback function for the /object_detection_emergency_stop topic.
        If this topic has a 'true' in it and we are on a waypoint mission,
        then we need to emergency stop the boat until the emergency stop is
        turned to 'false'.
        """
        if should_emergency_stop.data and self.motorboat_control_mode == MotorboatControlModes.WAYPOINT_MISSION:
            self.motorboat_control_mode = MotorboatControlModes.EMERGENCY_STOP

        elif not should_emergency_stop.data:
            self.motorboat_control_mode = MotorboatControlModes.WAYPOINT_MISSION




    def update_ros_topics(self) -> None:
        """
        This is the main function that is called constantely by the timer.
        Updates the propeller control struct and rudder angle topics based on the output of the autopilot controller.
        """

        desired_rudder_angle = 0.0
        desired_vesc_control_struct = VESCControlData(control_type_for_vesc="rpm", control_value=0.0)


        # Manually check whether any of the sensors have disconnecteds
        # if we have not received data from the sensor for 3 seconds then stop the propeller for safety reasons
        time_since_last_rc_data = time.time() - self.last_rc_data_received_time
        has_rc_data_disconnected = time_since_last_rc_data >= self.autopilot_parameters["rc_data_failsafe_time"]

        time_since_last_gps_data = time.time() - self.last_gps_position_received_time
        has_gps_disconnected = time_since_last_gps_data >= self.autopilot_parameters["gps_position_failsafe_time"]

        time_since_last_heading_data = time.time() - self.last_heading_received_time
        has_heading_disconnected = time_since_last_heading_data >= self.autopilot_parameters["heading_data_failsafe_time"]

        has_sensor_disconnected = has_rc_data_disconnected or has_gps_disconnected or has_heading_disconnected


        if has_sensor_disconnected:
            self.get_logger().info(f"has sensor disconnected: {has_sensor_disconnected}")
            self.get_logger().info(f"has remote controller disconnected: {has_rc_data_disconnected}")
            self.get_logger().info(f"has gps disconnected: {has_gps_disconnected}")
            self.get_logger().info(f"has heading disconnected: {has_heading_disconnected}")

            # Set the values to 0 as a failsafe so that the boat doesn't make bad decisions
            # while it lost connection with some of its sensors
            desired_vesc_control_struct = VESCControlData(control_type_for_vesc="rpm", control_value=0.0)

        elif self.motorboat_control_mode == MotorboatControlModes.EMERGENCY_STOP:
            desired_vesc_control_struct = VESCControlData(control_type_for_vesc="rpm", control_value=0.0)
            desired_rudder_angle = self.autopilot_parameters["rudder_hard_over"]

        elif self.motorboat_control_mode == MotorboatControlModes.WAYPOINT_MISSION and self.motorboat_autopilot.waypoints is not None:
            desired_rpm, desired_rudder_angle = self.motorboat_autopilot.run_waypoint_mission_step(self.position, self.heading)
            desired_vesc_control_struct = VESCControlData(control_type_for_vesc="rpm", control_value=desired_rpm)

        elif self.motorboat_control_mode == MotorboatControlModes.HOLD_HEADING:
            desired_rudder_angle = self.motorboat_autopilot.get_optimal_rudder_angle(self.heading, self.heading_to_hold)
            desired_vesc_control_type, desired_vesc_control_value, _ = self.motorboat_autopilot.run_rc_control(
                self.joystick_left_y, self.joystick_right_x, self.propeller_motor_control_mode
            )
            desired_vesc_control_struct = VESCControlData(
                control_type_for_vesc=desired_vesc_control_type, control_value=desired_vesc_control_value
            )

        elif self.motorboat_control_mode == MotorboatControlModes.FULL_RC:
            desired_vesc_control_type, desired_vesc_control_value, desired_rudder_angle = self.motorboat_autopilot.run_rc_control(
                self.joystick_left_y, self.joystick_right_x, self.propeller_motor_control_mode
            )
            desired_vesc_control_struct = VESCControlData(
                control_type_for_vesc=desired_vesc_control_type, control_value=desired_vesc_control_value
            )

        else:
            desired_vesc_control_struct = VESCControlData(control_type_for_vesc="rpm", control_value=0.0)


        # Now that we are done computing what the autopilot should do, publish everything to their respective topics

        self.propeller_motor_control_struct_publisher.publish(desired_vesc_control_struct)

        if desired_rudder_angle is not None:
            self.desired_rudder_angle_publisher.publish(Float32(data=float(desired_rudder_angle)))

        if self.should_zero_encoder:
            self.zero_rudder_encoder_publisher.publish(Bool(data=self.should_zero_encoder))
            self.encoder_has_been_zeroed = True

        self.should_propeller_motor_be_powered_publisher.publish(Bool(data=self.should_propeller_motor_be_powered))


        self.current_waypoint_index_publisher.publish(Int32(data=self.motorboat_autopilot.current_waypoint_index))

        self.boat_control_mode_publisher.publish(String(data=self.motorboat_control_mode.name))

        if self.motorboat_control_mode == MotorboatControlModes.HOLD_HEADING:
            self.desired_heading_publisher.publish(Float32(data=float(self.heading_to_hold)))

        elif self.motorboat_control_mode == MotorboatControlModes.WAYPOINT_MISSION and self.motorboat_autopilot.waypoints is not None:
            current_waypoint = self.motorboat_autopilot.waypoints[self.motorboat_autopilot.current_waypoint_index]
            bearing_to_waypoint = get_bearing(self.position, current_waypoint)
            self.desired_heading_publisher.publish(Float32(data=float(bearing_to_waypoint)))
        else:
            self.desired_heading_publisher.publish(Float32(data=0.))




def main() -> None:
    rclpy.init()
    motorboat_autopilot_node = MotorboatAutopilotNode()
    rclpy.spin(motorboat_autopilot_node)

    motorboat_autopilot_node.destroy_node()
    rclpy.shutdown()
